source/src/cip/cipstringi.c

```diff
@@ -154,3 +155,166 @@ void CipStringICopy(CipStringI *const to,
+					(CipString2* const ) target_stringI->array_of_string_i_structs[i].string;
+			CipUint length = GetUintFromMessage(&message_router_request->data);
+			SetCipString2ByData(string, length, message_router_request->data);
+			message_router_request->data += length * 2 * sizeof(CipOctet);
+		}
+			break;
+		case kCipStringN: {
+			CipUint size = GetUintFromMessage(&message_router_request->data);
+			CipUint length = GetUintFromMessage(&message_router_request->data);
+
+			target_stringI->array_of_string_i_structs[i].string = CipCalloc(1,
+					sizeof(CipStringN));
+			CipStringN *const string =
+					(CipStringN* const ) target_stringI->array_of_string_i_structs[i].string;
+			SetCipStringNByData(string, length, size,
+					message_router_request->data);
+			message_router_request->data += length * size;
+		}
+			break;
+		default:
+			OPENER_TRACE_ERR("CIP File: No valid String type received!\n");
+		}
+	} //end for
+}
+
+bool CipStringICompare(const CipStringI *const stringI_1,
+		const CipStringI *const stringI_2) {
+
+	/*loop through struct 1 strings*/
+	for (size_t i = 0; i < stringI_1->number_of_strings; ++i) {
+		// String 1
+		CipShortString *string_1 =
+				stringI_1->array_of_string_i_structs[i].string;
+		EipUint16 len_1 = string_1->length;
+
+		switch (stringI_1->array_of_string_i_structs[i].char_string_struct) {
+		case kCipShortString:
+			//default
+			break;
+		case kCipString: {
+			CipString *string_1 =
+					stringI_1->array_of_string_i_structs[i].string;
+			EipUint8 len_1 = string_1->length;
+		}
+			break;
+		case kCipString2: {
+			CipString2 *string_1 =
+					stringI_1->array_of_string_i_structs[i].string;
+			EipUint16 len_1 = string_1->length;
+		}
+			break;
+		case kCipStringN: {
+			CipStringN *string_1 =
+					stringI_1->array_of_string_i_structs[i].string;
+			EipUint16 length = string_1->length;
+			EipUint16 size = string_1->size; //bytes per symbol
+			len_1 = length * size;
+		}
+			break;
+		default:
+			OPENER_TRACE_ERR("CIP File: No valid String type received!\n");
+		}
+
+		/*loop through struct 2 strings*/
+		for (size_t i = 0; i < stringI_2->number_of_strings; ++i) {
+			// String 2
+			CipShortString *string_2 =
+					stringI_2->array_of_string_i_structs[i].string;
+			EipUint16 len_2 = string_2->length;
+
+			switch (stringI_2->array_of_string_i_structs[i].char_string_struct) {
+			case kCipShortString:
+				//default
+				break;
+			case kCipString: {
+				CipString *string_2 =
+						stringI_2->array_of_string_i_structs[i].string;
+				EipUint8 len_2 = string_2->length;
+			}
+				break;
+			case kCipString2: {
+				CipString2 *string_2 =
+						stringI_2->array_of_string_i_structs[i].string;
+				EipUint16 len_2 = string_2->length;
+			}
+				break;
+			case kCipStringN: {
+				CipStringN *string_2 =
+						stringI_2->array_of_string_i_structs[i].string;
+				EipUint16 length = string_2->length;
+				EipUint16 size = string_2->size; //bytes per symbol
+				len_2 = length * size;
+			}
```

Comment:
This cannot work imho. Have you created a test to verify the behavior?

You are creating new contexts with each {} in the cases, which also creates new variables called string_1 and string_2, shadowing the old variables, but for the memcmp only the uppermost are considered

Reply:
You need a void* pointer you reuse and remember which type of string needs to be compared and then cast appropriately

Reply:
Didn't hit the right review button, so bumping here